File: java-examples/scenario1-web-app/security-remediation/before/UserController.java
package com.enterprise.optimization.controller;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.web.bind.annotation.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import java.util.List;
import java.util.Map;

/**
 * REST controller for user operations.
 * This implementation contains several security vulnerabilities.
 */
@RestController
@RequestMapping("/api/users")
public class UserController {

    private static final Logger logger = LoggerFactory.getLogger(UserController.class);
    
    @Autowired
    private JdbcTemplate jdbcTemplate;

    /**
     * Get user by ID.
     * VULNERABILITY: SQL Injection due to string concatenation.
     */
    @GetMapping("/{id}")
    public Map<String, Object> getUserById(@PathVariable String id) {
        logger.info("Getting user with ID: " + id);
        
        // VULNERABLE: Direct string concatenation in SQL query
        String sql = "SELECT * FROM users WHERE user_id = " + id;
        
        List<Map<String, Object>> results = jdbcTemplate.queryForList(sql);
        if (results.isEmpty()) {
            return null;
        }
        return results.get(0);
    }

    /**
     * Search for users by name.
     * VULNERABILITY: SQL Injection in LIKE clause.
     */
    @GetMapping("/search")
    public List<Map<String, Object>> searchUsers(@RequestParam String name) {
        logger.info("Searching for users with name: " + name);
        
        // VULNERABLE: Unsanitized input in LIKE clause
        String sql = "SELECT * FROM users WHERE username LIKE '%" + name + "%'";
        
        return jdbcTemplate.queryForList(sql);
    }

    /**
     * Create a new user.
     * VULNERABILITY: No input validation, potential XSS and SQL Injection.
     */
    @PostMapping
    public Map<String, Object> createUser(@RequestBody User user) {
        logger.info("Creating new user: " + user.getUsername());
        
        // VULNERABLE: No input validation or sanitization
        String sql = "INSERT INTO users (username, email, password, role) VALUES ('"
                + user.getUsername() + "', '"
                + user.getEmail() + "', '"
                + user.getPassword() + "', '"
                + user.getRole() + "') RETURNING user_id";
        
        // VULNERABLE: Logging sensitive information
        logger.info("Creating user with password: " + user.getPassword());
        
        return jdbcTemplate.queryForMap(sql);
    }

    /**
     * Update user role.
     * VULNERABILITY: No authorization check.
     */
    @PutMapping("/{id}/role")
    public void updateUserRole(@PathVariable String id, @RequestParam String newRole) {
        logger.info("Updating role for user ID: " + id);
        
        // VULNERABLE: No authorization check for admin-only action
        String sql = "UPDATE users SET role = '" + newRole + "' WHERE user_id = " + id;
        
        jdbcTemplate.update(sql);
    }

    /**
     * Delete user by ID.
     * VULNERABILITY: No authorization check and SQL Injection.
     */
    @DeleteMapping("/{id}")
    public void deleteUser(@PathVariable String id) {
        logger.info("Deleting user with ID: " + id);
        
        // VULNERABLE: No authorization check and SQL Injection
        String sql = "DELETE FROM users WHERE user_id = " + id;
        
        jdbcTemplate.update(sql);
    }

    /**
     * Execute custom query (admin only).
     * VULNERABILITY: Direct execution of user input.
     */
    @PostMapping("/query")
    public List<Map<String, Object>> executeQuery(@RequestParam String query) {
        logger.info("Executing custom query");
        
        // VULNERABLE: Direct execution of user input
        return jdbcTemplate.queryForList(query);
    }

    /**
     * User class for request/response
     */
    public static class User {
        private String username;
        private String email;
        private String password;
        private String role;
        
        // Getters and setters
        public String getUsername() { return username; }
        public void setUsername(String username) { this.username = username; }
        
        public String getEmail() { return email; }
        public void setEmail(String email) { this.email = email; }
        
        public String getPassword() { return password; }
        public void setPassword(String password) { this.password = password; }
        
        public String getRole() { return role; }
        public void setRole(String role) { this.role = role; }
    }
} 

File: dotnet-examples/scenario1-web-app/security-remediation/before/UserController.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Data.SqlClient;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.Logging;

namespace Enterprise.Web.Controllers
{
    /// <summary>
    /// REST controller for user operations.
    /// This implementation contains several security vulnerabilities.
    /// </summary>
    [ApiController]
    [Route("api/[controller]")]
    public class UsersController : ControllerBase
    {
        private readonly ILogger<UsersController> _logger;
        private readonly string _connectionString;

        public UsersController(ILogger<UsersController> logger, IConfiguration configuration)
        {
            _logger = logger;
            _connectionString = configuration.GetConnectionString("DefaultConnection");
        }

        /// <summary>
        /// Get user by ID.
        /// VULNERABILITY: SQL Injection due to string concatenation.
        /// </summary>
        [HttpGet("{id}")]
        public async Task<IActionResult> GetUserById(string id)
        {
            _logger.LogInformation("Getting user with ID: " + id);
            
            // VULNERABLE: Direct string concatenation in SQL query
            string sql = "SELECT * FROM Users WHERE UserId = " + id;
            
            using (var connection = new SqlConnection(_connectionString))
            {
                await connection.OpenAsync();
                using (var command = new SqlCommand(sql, connection))
                {
                    using (var reader = await command.ExecuteReaderAsync())
                    {
                        if (await reader.ReadAsync())
                        {
                            var user = new User
                            {
                                UserId = reader["UserId"].ToString(),
                                Username = reader["Username"].ToString(),
                                Email = reader["Email"].ToString(),
                                Role = reader["Role"].ToString()
                            };
                            return Ok(user);
                        }
                    }
                }
            }
            
            return NotFound();
        }

        /// <summary>
        /// Search for users by name.
        /// VULNERABILITY: SQL Injection in LIKE clause.
        /// </summary>
        [HttpGet("search")]
        public async Task<IActionResult> SearchUsers([FromQuery] string name)
        {
            _logger.LogInformation("Searching for users with name: " + name);
            
            // VULNERABLE: Unsanitized input in LIKE clause
            string sql = "SELECT * FROM Users WHERE Username LIKE '%" + name + "%'";
            
            var users = new List<User>();
            using (var connection = new SqlConnection(_connectionString))
            {
                await connection.OpenAsync();
                using (var command = new SqlCommand(sql, connection))
                {
                    using (var reader = await command.ExecuteReaderAsync())
                    {
                        while (await reader.ReadAsync())
                        {
                            users.Add(new User
                            {
                                UserId = reader["UserId"].ToString(),
                                Username = reader["Username"].ToString(),
                                Email = reader["Email"].ToString(),
                                Role = reader["Role"].ToString()
                            });
                        }
                    }
                }
            }
            
            return Ok(users);
        }

        /// <summary>
        /// Create a new user.
        /// VULNERABILITY: No input validation, potential SQL Injection.
        /// </summary>
        [HttpPost]
        public async Task<IActionResult> CreateUser([FromBody] User user)
        {
            _logger.LogInformation("Creating new user: " + user.Username);
            
            // VULNERABLE: No input validation or sanitization
            string sql = $"INSERT INTO Users (Username, Email, Password, Role) VALUES ('{user.Username}', '{user.Email}', '{user.Password}', '{user.Role}'); SELECT SCOPE_IDENTITY()";
            
            // VULNERABLE: Logging sensitive information
            _logger.LogInformation("Creating user with password: " + user.Password);
            
            using (var connection = new SqlConnection(_connectionString))
            {
                await connection.OpenAsync();
                using (var command = new SqlCommand(sql, connection))
                {
                    var userId = await command.ExecuteScalarAsync();
                    user.UserId = userId.ToString();
                    return CreatedAtAction(nameof(GetUserById), new { id = user.UserId }, user);
                }
            }
        }

        /// <summary>
        /// Update user role.
        /// VULNERABILITY: No authorization check.
        /// </summary>
        [HttpPut("{id}/role")]
        public async Task<IActionResult> UpdateUserRole(string id, [FromQuery] string newRole)
        {
            _logger.LogInformation("Updating role for user ID: " + id);
            
            // VULNERABLE: No authorization check for admin-only action
            string sql = $"UPDATE Users SET Role = '{newRole}' WHERE UserId = {id}";
            
            using (var connection = new SqlConnection(_connectionString))
            {
                await connection.OpenAsync();
                using (var command = new SqlCommand(sql, connection))
                {
                    int rowsAffected = await command.ExecuteNonQueryAsync();
                    if (rowsAffected == 0)
                    {
                        return NotFound();
                    }
                    return NoContent();
                }
            }
        }

        /// <summary>
        /// Delete user by ID.
        /// VULNERABILITY: No authorization check and SQL Injection.
        /// </summary>
        [HttpDelete("{id}")]
        public async Task<IActionResult> DeleteUser(string id)
        {
            _logger.LogInformation("Deleting user with ID: " + id);
            
            // VULNERABLE: No authorization check and SQL Injection
            string sql = "DELETE FROM Users WHERE UserId = " + id;
            
            using (var connection = new SqlConnection(_connectionString))
            {
                await connection.OpenAsync();
                using (var command = new SqlCommand(sql, connection))
                {
                    int rowsAffected = await command.ExecuteNonQueryAsync();
                    if (rowsAffected == 0)
                    {
                        return NotFound();
                    }
                    return NoContent();
                }
            }
        }

        /// <summary>
        /// Execute custom query (admin only).
        /// VULNERABILITY: Direct execution of user input.
        /// </summary>
        [HttpPost("query")]
        public async Task<IActionResult> ExecuteQuery([FromQuery] string query)
        {
            _logger.LogInformation("Executing custom query");
            
            // VULNERABLE: Direct execution of user input
            var results = new List<Dictionary<string, object>>();
            
            using (var connection = new SqlConnection(_connectionString))
            {
                await connection.OpenAsync();
                using (var command = new SqlCommand(query, connection))
                {
                    using (var reader = await command.ExecuteReaderAsync())
                    {
                        while (await reader.ReadAsync())
                        {
                            var row = new Dictionary<string, object>();
                            for (int i = 0; i < reader.FieldCount; i++)
                            {
                                row[reader.GetName(i)] = reader.GetValue(i);
                            }
                            results.Add(row);
                        }
                    }
                }
            }
            
            return Ok(results);
        }
    }

    /// <summary>
    /// User class for request/response
    /// </summary>
    public class User
    {
        public string UserId { get; set; }
        public string Username { get; set; }
        public string Email { get; set; }
        public string Password { get; set; }
        public string Role { get; set; }
    }
} 